Makefile

```diff
@@ -0,0 +1,40 @@
+CXX          ?= g++
+CXXOPTS      ?= -g -Wall -Werror -std=c++11 -Iinclude/
+DEBUG_OPTS   ?= -fprofile-arcs -ftest-coverage -O0 -g -fPIC
+RELEASE_OPTS ?= -O3
+
+all: test
+
+# Release libraries
+release/libsimhash.o: release/simhash.o release/permutation.o
+	mkdir -p release
+	ld -r -o $@ $^
+
+release/%.o: src/%.cpp include/%.h
+	mkdir -p release
+	$(CXX) $(CXXOPTS) $(RELEASE_OPTS) -o $@ -c $<
+
+# Debug libraries
+debug/libsimhash.o: debug/simhash.o debug/permutation.o
+	mkdir -p debug
+	ld -r -o $@ $^
+
+debug/%.o: src/%.cpp include/%.h
+	mkdir -p debug
+	$(CXX) $(CXXOPTS) $(DEBUG_OPTS) -o $@ -c $<
+
+test/%.o: test/%.cpp
+	$(CXX) $(CXXOPTS) $(DEBUG_OPTS) -o $@ -c $<
+
+# Tests
+test-all: test/test-all.o test/test-simhash.o test/test-permutation.o debug/libsimhash.o
+	$(CXX) $(CXXOPTS) $(DEBUG_OPTS) -o $@ $^ -lgtest -lpthread
+
+.PHONY: test
+test: test-all
+	./test-all
+
+clean:
+	find . -name '*.o' -o -name '*.gcda' -o -name '*.gcno' -o -name '*.gcov' \
```

Comment:
I would probably just change this to `rm -rf debug release` assuming all of your build artifacts land there.